Gems/Terrain/Code/Source/TerrainRenderer/TerrainMeshManager.cpp

```diff
@@ -388,39 +415,247 @@ namespace Terrain
+
+        if (numSamples.first != paddedVertexCount1d || numSamples.second != paddedVertexCount1d)
+        {
+            AZ_Assert(false, "Number of samples returned from GetNumSamplesFromRegion does not match expectations");
+            return {};
+        }
+
+        uint16_t meshVertexCount = vertexCount1d * vertexCount1d;
+        uint16_t queryVertexCount = paddedVertexCount1d * paddedVertexCount1d;
+        AZStd::vector<float> heights;
+        heights.resize_no_construct(queryVertexCount);
+
+        AZStd::vector<uint16_t> meshHeights;
+        AZStd::vector<AZStd::pair<uint16_t, uint16_t>> meshNormals;
+        meshHeights.resize_no_construct(meshVertexCount);
+        meshNormals.resize_no_construct(meshVertexCount);
+
+        auto perPositionCallback = [this, &heights, paddedVertexCount1d]
+        ([[maybe_unused]] size_t xIndex, [[maybe_unused]] size_t yIndex,
+            const AzFramework::SurfaceData::SurfacePoint& surfacePoint,
+            [[maybe_unused]] bool terrainExists)
+        {
+            const float height = surfacePoint.m_position.GetZ() - m_worldBounds.GetMin().GetZ();
+            heights.at(yIndex * paddedVertexCount1d + xIndex) = height;
+        };
+
+        AzFramework::Terrain::TerrainDataRequestBus::Broadcast(
+            &AzFramework::Terrain::TerrainDataRequests::ProcessHeightsFromRegion,
+            queryBounds, stepSize, perPositionCallback, samplerType);
+
+        float rcpWorldZ = 1.0f / m_worldBounds.GetExtents().GetZ();
+        float vertexSpacing2 = vertexSpacing * 2.0f;
+        float minHeight = AZStd::numeric_limits<float>::max();
+        float maxHeight = AZStd::numeric_limits<float>::min();
+
+        // float versions of int max to make sure a int->float conversion doesn't happen at each loop iteration.
+        constexpr float maxUint16 = AZStd::numeric_limits<uint16_t>::max();
+        constexpr float maxInt16 = AZStd::numeric_limits<int16_t>::max();
+
+        for (uint16_t y = 0; y < vertexCount1d; ++y)
+        {
+            const uint16_t offsetY = y + 1;
+
+            for (uint16_t x = 0; x < vertexCount1d; ++x)
+            {
+                const uint16_t offsetX = x + 1;
+                const uint16_t offsetCoord = offsetY * paddedVertexCount1d + offsetX;
+                const uint16_t coord = y * vertexCount1d + x;
+
+                const float height = heights.at(offsetCoord);
+                const float clampedHeight = AZ::GetClamp(height * rcpWorldZ, 0.0f, 1.0f);
+                const uint16_t uint16Height = aznumeric_cast<uint16_t>(clampedHeight * maxUint16 + 0.5f); // always positive, so just add 0.5 to round.
+                meshHeights.at(coord) = uint16Height;
+
+                if (minHeight > height)
+                {
+                    minHeight = height;
+                }
+                else if (maxHeight < height)
+                {
+                    maxHeight = height;
+                }
```

Comment:
Current code is slightly faster because it can avoid the second operation. There is a bug though in that It does require a better initial state so I'll initialize max and min height to the first height value.